Sources/Fuzzilli/Core/CodeTemplate.swift

```diff
@@ -0,0 +1,114 @@
+        f(b)
+    }
+
+    /// Generate an array of random function signatures
+    static func generateRandomFunctionSignatures(forFuzzer fuzzer: Fuzzer, n: Int) -> [FunctionSignature] {
+        var signatures = [FunctionSignature]()
+        for _ in 0..<n {
+            signatures.append(CodeTemplate.generateSignature(forFuzzer: fuzzer, n: Int.random(in: 0...3)))
+        }
+        return signatures
+    }
+
+    /// This function generates and sets property types for the global properties
+    static func generateRandomPropertyTypes(forBuilder b: ProgramBuilder) {
+        // generate types for half of the available property names.
+        for i in 0..<b.fuzzer.environment.customPropertyNames.count/2 {
+            // if we pass "e" (last customPropertyName) into generateType the filter might fail and we
+            // end up not setting a type for this property. Therefore just take the next one if it is "e".
+            // TODO: fix this
+            var name = Array(b.fuzzer.environment.customPropertyNames)[i]
+            if name == "e" {
+                name = Array(b.fuzzer.environment.customPropertyNames)[i+1]
+            }
+            b.setType(ofProperty: name, to: CodeTemplate.generateType(forFuzzer: b.fuzzer, forProperty: name))
+        }
+    }
+
+    /// Generate and set random method types for global method names.
+    static func generateRandomMethodTypes(forBuilder b: ProgramBuilder, n: Int) {
+        for _ in 0..<n {
+            let method = chooseUniform(from: b.fuzzer.environment.methodNames)
+            let signature = CodeTemplate.generateSignature(forFuzzer: b.fuzzer, n: Int.random(in: 0..<3))
+            b.setSignature(ofMethod: method, to: signature)
+        }
+    }
+
+    /// Generate a random type to use in e.g. function signatures.
+    /// This function should only emit types that can be constructed by ProgramBuilder.generateVariable.
+    static func generateType(forFuzzer fuzzer: Fuzzer, forProperty property: String = "") -> Type {
+        return withEqualProbability(
+            // Choose a basic type
+            { () -> Type in
+                chooseUniform(from: [.integer, .float, .boolean, .bigint, .string])
+            },
+            // Choose an array
+            {
+                return .object(ofGroup: "Array")
+            },
+            // choose a complicated object
+            {
+                var properties: [String] = []
+                var methods: [String] = []
+
+                // TODO: add group with probability here as well?
+
+                // Generate random properties.
+                // We filter the candidates to avoid cycles in our objects.
+                // TODO: we should remove this "no-cycle" restriction here, and let `generateVariable`
+                // handle these cases.
+                for _ in 1..<3 {
+                    let candidates = fuzzer.environment.customPropertyNames.filter({ $0 >= property })
```

Comment:
Shouldn't this be `>` instead of `>=`?
Otherwise I think you can end up with
`.b = Type.object(withProperties: ["b"])` 
and I think then generateVariable would recurse until it crashes.